lib/puppet/provider/openldap.rb

```diff
@@ -0,0 +1,134 @@
+require 'tempfile'
+
+class Puppet::Provider::Openldap < Puppet::Provider
+
+  initvars # without this, commands won't work
+
+  commands :original_slapcat    => 'slapcat',
+           :original_ldapmodify => 'ldapmodify',
+           :original_ldapadd    => 'ldapadd'
+
+  def self.slapcat(filter, dn = '', base = 'cn=config')
+    arguments = [
+      '-b', base,
+      '-o', 'ldif-wrap=no',
+      '-H', "ldap:///#{dn}???#{filter}"
+    ]
+
+    if 'squeeze' == Facter.value(:lsbdistcodename)
+      arguments = [
+        '-b', base,
+        '-H', "ldap:///#{dn}???#{filter}"
+      ]
+    end
+
+    original_slapcat(*arguments)
+  end
+  def slapcat(*args); self.class.slapcat(*args); end
+
+  def self.ldapadd(path)
+    original_ldapadd('-cQY', 'EXTERNAL', '-H', 'ldapi:///', '-f', path)
+  end
+  def ldapadd(*args); self.class.ldapadd(*args); end
+
+  # Unwrap LDIF and return each attribute beginning with "olc" also removing
+  # that occurance of "olc" at the beginning.
+  def self.get_lines(items)
+    items.strip.
+      gsub("\n ", "").
+      split("\n").
+      select  { |entry| entry =~ /^olc/ }.
+      collect { |entry| entry.gsub(/^olc/, '') }
+  end
+  def get_lines(*args); self.class.get_lines(*args); end
+
+  # Unwrap LDIF and return each entry as array of lines.
+  #
+  # Example LDIF:
+  #   dn: cn=config
+  #   ...
+  #
+  #   dn: cn=schema,cn=config
+  #   ...
+  #
+  # Results in:
+  #
+  #   [['dn: cn=config', '...'],
+  #    ['dn: cn=schema,cn=config', '...']]
+  #
+  def self.get_entries(items)
+    items.strip.
+      split("\n\n").
+      collect { |paragraph|
+        paragraph.
+          gsub("\n ", '').
+          split("\n")
+      }
+  end
+  def get_entries(*args); self.class.get_entries(*args); end
+
+  def self.last_of_split(line, by = ' ')
+    line.split(by, 2).last
+  end
+  def last_of_split(*args); self.class.last_of_split(*args); end
+
+  def self.ldapmodify(path)
+    original_ldapmodify('-Y', 'EXTERNAL', '-H', 'ldapi:///', '-f', path)
+  end
+  def ldapmodify(*args); self.class.ldapmodify(*args); end
```

Comment:
What do you mean? Both methods are just there to ensure the availability in class and "instance" methods. (self.instances, create, etc.)